### ISSUE TITLE
S3 multipart upload always fails checksum with `--s3-no-head` option

### ISSUE THREAD

**Issue by c42f** [rclone/rclone#5956]
When the `--s3-no-head` option is set and when a multipart upload is attempted (by default at the 200MiB boundary), `rclone copyto` fails reproducibly with a checksum error.

Full details and `rclone` command are in the log below, but the relevant part of the log seems to be:

```
corrupted on transfer: md5 hash differ "100a7ce0ac810f7d1caa75ab9b5d3bfc" vs "EAp84KyBD30cqnWrm107/A=="
```

`"EAp84KyBD30cqnWrm107/A=="` doesn't even look like a MD5 checksum so what's going on?  This seems to be the base64 encoded version of the correct checksum, as can be verified with a quick Julia script:

```julia
julia> using Base64

julia> bytes2hex(base64decode("EAp84KyBD30cqnWrm107/A=="))
"100a7ce0ac810f7d1caa75ab9b5d3bfc"

julia> bytes2hex(base64decode("EAp84KyBD30cqnWrm107/A=="))  ==  "100a7ce0ac810f7d1caa75ab9b5d3bfc"
true
```

With the `--s3-no-head` option removed, this error disappears.

Here's the detailed bug report form filled out:

#### What is your rclone version (output from `rclone version`)

```
rclone v1.57.0
- os/version: ubuntu 18.04 (64 bit)
- os/kernel: 4.15.0-166-generic (x86_64)
- os/type: linux
- os/arch: amd64
- go/version: go1.17.2
- go/linking: static
- go/tags: none
```

#### Which OS you are using and how many bits (e.g. Windows 7, 64 bit)

Ubuntu 18.04, 64 bit

#### Which cloud storage system are you using? (e.g. Google Drive)

AWS S3

#### The command you were trying to run (e.g. `rclone copy /tmp remote:tmp`)

I've reduced the `--s3-upload-cutoff` and `--s3-chunk-size` here for ease of testing, and only upload a 10MB file.  But removing these options and using a 210MB file will also display the bug.

```bash
head -c 10M </dev/urandom > myfile

/home/chris/local/rclone-v1.57.0-linux-amd64/rclone copyto myfile \
    jc:datasets-jl-test/rand_10M.dat \
    --config rclone.conf \
    --s3-no-head \
    --s3-upload-cutoff 1M --s3-chunk-size 5M
```

#### A log from the command with the `-vv` flag (e.g. output from `rclone -vv copy /tmp remote:tmp`)

```
2022/01/27 14:08:54 DEBUG : rclone: Version "v1.57.0" starting with parameters ["/home/chris/local/rclone-v1.57.0-linux-amd64/rclone" "copyto" "myfile" "jc:datasets-jl-test/rand_10M.dat" "--config" "rclone.conf" "--s3-no-head" "--s3-upload-cutoff" "1M" "--s3-chunk-size" "5M" "-vv" "--log-file" "bug.log"]
2022/01/27 14:08:54 DEBUG : Creating backend with remote "myfile"
2022/01/27 14:08:54 DEBUG : Using config file from "/home/chris/work/ops/4171_rclone_multipart_checksum/rclone.conf"
2022/01/27 14:08:54 DEBUG : fs cache: adding new entry for parent of "myfile", "/home/chris/work/ops/4171_rclone_multipart_checksum"
2022/01/27 14:08:54 DEBUG : Creating backend with remote "jc:datasets-jl-test/"
2022/01/27 14:08:54 DEBUG : jc: detected overridden config - adding "{QX2bg}" suffix to name
2022/01/27 14:08:54 DEBUG : fs cache: renaming cache item "jc:datasets-jl-test/" to be canonical "jc{QX2bg}:datasets-jl-test"
2022/01/27 14:08:54 DEBUG : myfile: Need to transfer - File not found at Destination
2022/01/27 14:08:55 DEBUG : rand_10M.dat: multipart upload starting chunk 1 size 5Mi offset 0/10Mi
2022/01/27 14:08:55 DEBUG : rand_10M.dat: multipart upload starting chunk 2 size 5Mi offset 5Mi/10Mi
2022/01/27 14:09:00 DEBUG : myfile: md5 = 100a7ce0ac810f7d1caa75ab9b5d3bfc (Local file system at /home/chris/work/ops/4171_rclone_multipart_checksum)
2022/01/27 14:09:00 DEBUG : rand_10M.dat: md5 = EAp84KyBD30cqnWrm107/A== (S3 bucket datasets-jl-test)
2022/01/27 14:09:00 ERROR : rand_10M.dat: corrupted on transfer: md5 hash differ "100a7ce0ac810f7d1caa75ab9b5d3bfc" vs "EAp84KyBD30cqnWrm107/A=="
2022/01/27 14:09:00 INFO  : rand_10M.dat: Removing failed copy
2022/01/27 14:09:00 ERROR : Attempt 1/3 failed with 1 errors and: corrupted on transfer: md5 hash differ "100a7ce0ac810f7d1caa75ab9b5d3bfc" vs "EAp84KyBD30cqnWrm107/A=="
2022/01/27 14:09:00 DEBUG : myfile: Need to transfer - File not found at Destination
2022/01/27 14:09:00 DEBUG : rand_10M.dat: multipart upload starting chunk 1 size 5Mi offset 0/10Mi
2022/01/27 14:09:00 DEBUG : rand_10M.dat: multipart upload starting chunk 2 size 5Mi offset 5Mi/10Mi
2022/01/27 14:09:05 DEBUG : myfile: md5 = 100a7ce0ac810f7d1caa75ab9b5d3bfc (Local file system at /home/chris/work/ops/4171_rclone_multipart_checksum)
2022/01/27 14:09:05 DEBUG : rand_10M.dat: md5 = EAp84KyBD30cqnWrm107/A== (S3 bucket datasets-jl-test)
2022/01/27 14:09:05 ERROR : rand_10M.dat: corrupted on transfer: md5 hash differ "100a7ce0ac810f7d1caa75ab9b5d3bfc" vs "EAp84KyBD30cqnWrm107/A=="
2022/01/27 14:09:05 INFO  : rand_10M.dat: Removing failed copy
2022/01/27 14:09:05 ERROR : Attempt 2/3 failed with 1 errors and: corrupted on transfer: md5 hash differ "100a7ce0ac810f7d1caa75ab9b5d3bfc" vs "EAp84KyBD30cqnWrm107/A=="
2022/01/27 14:09:05 DEBUG : myfile: Need to transfer - File not found at Destination
2022/01/27 14:09:05 DEBUG : rand_10M.dat: multipart upload starting chunk 1 size 5Mi offset 0/10Mi
2022/01/27 14:09:05 DEBUG : rand_10M.dat: multipart upload starting chunk 2 size 5Mi offset 5Mi/10Mi
2022/01/27 14:09:10 DEBUG : myfile: md5 = 100a7ce0ac810f7d1caa75ab9b5d3bfc (Local file system at /home/chris/work/ops/4171_rclone_multipart_checksum)
2022/01/27 14:09:10 DEBUG : rand_10M.dat: md5 = EAp84KyBD30cqnWrm107/A== (S3 bucket datasets-jl-test)
2022/01/27 14:09:10 ERROR : rand_10M.dat: corrupted on transfer: md5 hash differ "100a7ce0ac810f7d1caa75ab9b5d3bfc" vs "EAp84KyBD30cqnWrm107/A=="
2022/01/27 14:09:10 INFO  : rand_10M.dat: Removing failed copy
2022/01/27 14:09:10 ERROR : Attempt 3/3 failed with 1 errors and: corrupted on transfer: md5 hash differ "100a7ce0ac810f7d1caa75ab9b5d3bfc" vs "EAp84KyBD30cqnWrm107/A=="
2022/01/27 14:09:10 INFO  : 
Transferred:   	       30 MiB / 30 MiB, 100%, 1.875 MiB/s, ETA 0s
Errors:                 1 (retrying may help)
Elapsed time:        16.5s

2022/01/27 14:09:10 DEBUG : 9 go routines active
2022/01/27 14:09:10 Failed to copyto: corrupted on transfer: md5 hash differ "100a7ce0ac810f7d1caa75ab9b5d3bfc" vs "EAp84KyBD30cqnWrm107/A=="
```

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by ncw** [comment]
This looks easy to fix!

Have a go with this

[v1.58.0-beta.5977.57d80cfb5.fix-5956-s3-no-head-upload](https://beta.rclone.org/branch/fix-5956-s3-no-head-upload/v1.58.0-beta.5977.57d80cfb5.fix-5956-s3-no-head-upload/) on branch [fix-5956-s3-no-head-upload](https://github.com/rclone/rclone/tree/fix-5956-s3-no-head-upload) (uploaded in 15-30 mins)

**Comment by c42f** [comment]
Great! I can confirm this works with my test case.

**Comment by ncw** [comment]
Thanks for testing.

I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.58
